### PR TITLE
[V2] fix: only count replica attachments w/t deletion timestamp from cache client

### PR DESCRIPTION
**Pull request by hccheng72**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Because the update delay of cache client, some deleted CRIs still can be found, and the driver gets wrong data to process. For example, when a replica attachment failed and get deleted to restart the attach request, if the driver still found the deleted CRI from cache, we will get a smaller number of replica attachments than that we actually need to create. 

**Which issue(s) this PR fixes**:

Fixes #

**Requirements**:
- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Special notes for your reviewer**:

**Release note**:
```
none
```

### CODE REVIEW

File: pkg/controller/shared_state.go [hunk 1]
```diff
@@ -1258,7 +1260,14 @@ func (c *SharedState) manageReplicas(ctx context.Context, volumeName string) err
 		return err
 	}
 
-	desiredReplicaCount, currentReplicaCount := azVolume.Spec.MaxMountReplicaCount, len(azVolumeAttachments)
+	currentReplicaCount := 0
+	for _, azVolumeAttachment := range azVolumeAttachments {
```
[the remark this answers is not in the document]
Comment (hccheng72) [reply pkg/controller/shared_state.go:1264]: I see `getNodesWithReplica()` also call `GetAzVolumeAttachmentsForVolume` and `for` loop in its implementation but we don't need the list of nodes.

[the remark this answers is not in the document]
Comment (hccheng72) [reply pkg/controller/shared_state.go:1264]: I see. The benefit of using `getNodesWithReplica` will make the code much cleaner but also use additional memory to store the list of nodes. I have no preference for either one. So yeah, we can replace this with `getNodesWithReplica`.

[the remark this answers is not in the document]
Comment (landreasyan) [reply pkg/controller/shared_state.go:1264]: I believe using `GetAzVolumeAttachmentsForVolume` makes the intend of our code and readability much better. For understanding the usage of `getNodesWithReplica` one must also know that there is one replica per node.

[the remark this answers is not in the document]
Comment (hccheng72) [reply pkg/controller/shared_state.go:1264]: Sounds good for me!

**Review by sunpa93 — APPROVED** [review]
/lgtm

File: pkg/controller/shared_state.go [hunk 2]
```diff
@@ -1252,13 +1254,12 @@ func (c *SharedState) manageReplicas(ctx context.Context, volumeName string) err
 		return nil
 	}
 
-	azVolumeAttachments, err := azureutils.GetAzVolumeAttachmentsForVolume(ctx, c.cachedClient, volumeName, azureutils.ReplicaOnly)
-	if err != nil {
-		w.Logger().Errorf(err, "failed to list replica AzVolumeAttachments")
+	currentReplicaCount := 0
+	if currentReplicaCount, err = c.countValidReplicasForVolume(ctx, volumeName); err != nil {
```
Comment (landreasyan) [inline pkg/controller/shared_state.go:1258]: np: move `currentReplicaCount, err = c.countValidReplicasForVolume(ctx, volumeName)` outside of the if scope to eliminate the unnecessary initialization of `currentReplicaCount:=0`

Comment (hccheng72) [reply pkg/controller/shared_state.go:1258]: But we use the same `err` variable in this method for defer function `w.Finish(err)`. I still need to create and initialize `currentReplicaCount` variable first even if `countValidReplicasForVolume` is moved outside of `if` scope. Is it right?

Comment (hccheng72) [reply pkg/controller/shared_state.go:1258]: You're right. Create a new `err` variable shouldn't affect the workflow to call `Finish(err)`.

**Pull request merged**